File: sdks/go/client_test.go
package notifyd

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// newTestServer builds an httptest server that issues token
// "test-token-<n>" on each /auth/token call (n increments per call, so
// tests can distinguish a cached token from a freshly-issued one) and
// dispatches every other path through handler. handler receives the
// Authorization header's bearer token so tests can assert on it. The
// returned tokenCallCount function is safe to call concurrently with
// in-flight requests.
func newTestServer(t *testing.T, handler func(w http.ResponseWriter, r *http.Request, bearerToken string)) (server *httptest.Server, client *Client, tokenCallCount func() int64) {
	t.Helper()

	var callCount int64
	mux := http.NewServeMux()
	mux.HandleFunc("/auth/token", func(w http.ResponseWriter, r *http.Request) {
		n := atomic.AddInt64(&callCount, 1)
		var req map[string]string
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatalf("token request body: %v", err)
		}
		if req["api_key"] != "test-key" || req["api_secret"] != "test-secret" {
			w.WriteHeader(http.StatusUnauthorized)
			json.NewEncoder(w).Encode(APIError{Error: "INVALID_CREDENTIALS"})
			return
		}
		json.NewEncoder(w).Encode(tokenResponse{
			Token:     fmt.Sprintf("test-token-%d", n),
			ExpiresIn: "24h0m0s",
		})
	})
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		auth := r.Header.Get("Authorization")
		const prefix = "Bearer "
		token := ""
		if len(auth) > len(prefix) {
			token = auth[len(prefix):]
		}
		handler(w, r, token)
	})

	testServer := httptest.NewServer(mux)
	t.Cleanup(testServer.Close)

	newClient, err := New(Config{
		APIKey:    "test-key",
		APISecret: "test-secret",
		BaseURL:   testServer.URL,
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return testServer, newClient, func() int64 { return atomic.LoadInt64(&callCount) }
}

func TestTokenExchangeAndCaching(t *testing.T) {
	ctx := context.Background()
	var channelsCalls int
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/channels" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		channelsCalls++
		if token != "test-token-1" {
			t.Errorf("call %d: got token %q, want test-token-1 (should reuse cached token)", channelsCalls, token)
		}
		json.NewEncoder(w).Encode([]ChannelConfig{})
	})

	if _, err := client.ListChannels(ctx); err != nil {
		t.Fatalf("first ListChannels: %v", err)
	}
	if _, err := client.ListChannels(ctx); err != nil {
		t.Fatalf("second ListChannels: %v", err)
	}
	if channelsCalls != 2 {
		t.Fatalf("got %d channels calls, want 2", channelsCalls)
	}
}

func TestRefreshOnceOn401(t *testing.T) {
	ctx := context.Background()
	var requestTokens []string
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		requestTokens = append(requestTokens, token)
		// Reject the first token to simulate an expired/revoked token the
		// client didn't know about yet; accept any subsequent token.
		if token == "test-token-1" {
			w.WriteHeader(http.StatusUnauthorized)
			json.NewEncoder(w).Encode(APIError{Error: "TOKEN_EXPIRED"})
			return
		}
		json.NewEncoder(w).Encode([]ChannelConfig{})
	})

	if _, err := client.ListChannels(ctx); err != nil {
		t.Fatalf("ListChannels: %v", err)
	}
	if len(requestTokens) != 2 {
		t.Fatalf("got %d requests, want 2 (initial + one retry)", len(requestTokens))
	}
	if requestTokens[0] != "test-token-1" || requestTokens[1] != "test-token-2" {
		t.Fatalf("got tokens %v, want [test-token-1 test-token-2]", requestTokens)
	}
}

func TestRefreshOnlyOnceNotLooped(t *testing.T) {
	ctx := context.Background()
	var attempts int
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		attempts++
		// Every attempt returns 401, simulating persistently bad
		// credentials. The client must give up after one retry, not loop.
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(APIError{Error: "INVALID_TOKEN"})
	})

	_, err := client.ListChannels(ctx)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	reqErr, ok := err.(*RequestError)
	if !ok {
		t.Fatalf("got error type %T, want *RequestError", err)
	}
	if reqErr.StatusCode != http.StatusUnauthorized || reqErr.Code != "INVALID_TOKEN" {
		t.Fatalf("got %+v, want 401 INVALID_TOKEN", reqErr)
	}
	if attempts != 2 {
		t.Fatalf("got %d attempts, want 2 (initial + one retry, no loop)", attempts)
	}
}

func TestSend(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/notifications/send" || r.Method != http.MethodPost {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body SendInput
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		if body.ChannelConfigID != "chan-1" || body.Body != "hello" {
			t.Fatalf("unexpected body: %+v", body)
		}
		w.WriteHeader(http.StatusAccepted)
		json.NewEncoder(w).Encode(Notification{ID: "notif-1", Status: StatusPending})
	})

	notification, err := client.Send(ctx, SendInput{ChannelConfigID: "chan-1", Body: "hello"})
	if err != nil {
		t.Fatalf("Send: %v", err)
	}
	if notification.ID != "notif-1" || notification.Status != StatusPending {
		t.Fatalf("got %+v", notification)
	}
}

func TestSend_QuotaExceededExposesUpgradeURL(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		w.WriteHeader(http.StatusTooManyRequests)
		json.NewEncoder(w).Encode(map[string]any{
			"error":       "QUOTA_EXCEEDED",
			"upgrade_url": "https://notifyd.fluxintek.com/account/upgrade",
		})
	})

	_, err := client.Send(ctx, SendInput{ChannelConfigID: "chan-1", Body: "hello"})
	if err == nil {
		t.Fatal("expected error, got nil")
	}

	reqErr, ok := err.(*RequestError)
	if !ok {
		t.Fatalf("got error type %T, want *RequestError", err)
	}
	if reqErr.StatusCode != http.StatusTooManyRequests || reqErr.Code != "QUOTA_EXCEEDED" {
		t.Fatalf("got %+v, want 429 QUOTA_EXCEEDED", reqErr)
	}

	upgradeURL, ok := reqErr.Field("upgrade_url")
	if !ok {
		t.Fatalf("expected upgrade_url field to be present in %+v", reqErr.Body)
	}
	if upgradeURL != "https://notifyd.fluxintek.com/account/upgrade" {
		t.Fatalf("got upgrade_url %q, want the account upgrade link", upgradeURL)
	}
}

func TestSendMulti(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/notifications/send-multi" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		w.WriteHeader(http.StatusAccepted)
		json.NewEncoder(w).Encode(SendMultiResponse{
			Sent:   []Notification{{ID: "notif-1"}},
			Errors: []string{"chan-2: CHANNEL_NOT_FOUND"},
		})
	})

	result, err := client.SendMulti(ctx, []SendInput{
		{ChannelConfigID: "chan-1", Body: "a"},
		{ChannelConfigID: "chan-2", Body: "b"},
	})
	if err != nil {
		t.Fatalf("SendMulti: %v", err)
	}
	if len(result.Sent) != 1 || len(result.Errors) != 1 {
		t.Fatalf("got %+v", result)
	}
}

func TestListNotifications(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/notifications" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if got := r.URL.Query().Get("status"); got != "delivered" {
			t.Fatalf("got status filter %q, want delivered", got)
		}
		json.NewEncoder(w).Encode(NotificationList{
			Data:  []Notification{{ID: "notif-1", Status: StatusDelivered}},
			Total: 1, Limit: 20, Offset: 0,
		})
	})

	list, err := client.ListNotifications(ctx, ListNotificationsParams{Status: StatusDelivered})
	if err != nil {
		t.Fatalf("ListNotifications: %v", err)
	}
	if list.Total != 1 || len(list.Data) != 1 {
		t.Fatalf("got %+v", list)
	}
}

func TestGetNotification(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/notifications/notif-1" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		json.NewEncoder(w).Encode(Notification{ID: "notif-1", Status: StatusDelivered})
	})

	notification, err := client.GetNotification(ctx, "notif-1")
	if err != nil {
		t.Fatalf("GetNotification: %v", err)
	}
	if notification.Status != StatusDelivered {
		t.Fatalf("got %+v", notification)
	}
}

func TestListAttempts(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		if r.URL.Path != "/notifications/notif-1/attempts" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		json.NewEncoder(w).Encode([]DeliveryAttempt{{ID: "attempt-1", AttemptNumber: 1, Status: AttemptSuccess}})
	})

	attempts, err := client.ListAttempts(ctx, "notif-1")
	if err != nil {
		t.Fatalf("ListAttempts: %v", err)
	}
	if len(attempts) != 1 || attempts[0].Status != AttemptSuccess {
		t.Fatalf("got %+v", attempts)
	}
}

func TestChannelsCRUD(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/channels":
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(ChannelConfig{ID: "chan-1", Channel: ChannelTelegram})
		case r.Method == http.MethodGet && r.URL.Path == "/channels/chan-1":
			json.NewEncoder(w).Encode(ChannelConfig{ID: "chan-1", Channel: ChannelTelegram})
		case r.Method == http.MethodPatch && r.URL.Path == "/channels/chan-1":
			json.NewEncoder(w).Encode(ChannelConfig{ID: "chan-1", Channel: ChannelTelegram, IsActive: false})
		case r.Method == http.MethodDelete && r.URL.Path == "/channels/chan-1":
			w.WriteHeader(http.StatusNoContent)
		default:
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
	})

	created, err := client.CreateChannel(ctx, CreateChannelInput{
		Channel: ChannelTelegram,
		Name:    "ops-alerts",
		Config:  map[string]any{"bot_token": "x", "chat_id": "y"},
	})
	if err != nil || created.ID != "chan-1" {
		t.Fatalf("CreateChannel: %+v, %v", created, err)
	}

	fetched, err := client.GetChannel(ctx, "chan-1")
	if err != nil || fetched.ID != "chan-1" {
		t.Fatalf("GetChannel: %+v, %v", fetched, err)
	}

	inactive := false
	updated, err := client.UpdateChannel(ctx, "chan-1", UpdateChannelInput{IsActive: &inactive})
	if err != nil || updated.IsActive {
		t.Fatalf("UpdateChannel: %+v, %v", updated, err)
	}

	if err := client.DeleteChannel(ctx, "chan-1"); err != nil {
		t.Fatalf("DeleteChannel: %v", err)
	}
}

func TestKeysCRUD(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		switch {
		case r.Method == http.MethodGet && r.URL.Path == "/keys":
			json.NewEncoder(w).Encode([]APIKey{{ID: "key-1", Label: "ci"}})
		case r.Method == http.MethodPost && r.URL.Path == "/keys":
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(CreateAPIKeyResponse{
				Key:       APIKey{ID: "key-2", Label: "new"},
				APISecret: "shown-once",
			})
		case r.Method == http.MethodDelete && r.URL.Path == "/keys/key-1":
			w.WriteHeader(http.StatusNoContent)
		default:
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
	})

	keys, err := client.ListAPIKeys(ctx)
	if err != nil || len(keys) != 1 {
		t.Fatalf("ListAPIKeys: %+v, %v", keys, err)
	}

	created, err := client.CreateAPIKey(ctx, "new")
	if err != nil || created.APISecret != "shown-once" {
		t.Fatalf("CreateAPIKey: %+v, %v", created, err)
	}

	if err := client.RevokeAPIKey(ctx, "key-1"); err != nil {
		t.Fatalf("RevokeAPIKey: %v", err)
	}
}

func TestWebhooksCRUD(t *testing.T) {
	ctx := context.Background()
	_, client, _ := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		switch {
		case r.Method == http.MethodGet && r.URL.Path == "/webhooks":
			json.NewEncoder(w).Encode([]WebhookEndpoint{{ID: "wh-1"}})
		case r.Method == http.MethodPost && r.URL.Path == "/webhooks":
			w.WriteHeader(http.StatusCreated)
			json.NewEncoder(w).Encode(WebhookEndpointCreated{
				WebhookEndpoint: WebhookEndpoint{ID: "wh-2", URL: "https://example.com/hook"},
				Secret:          "whsec_shown_once",
			})
		case r.Method == http.MethodPut && r.URL.Path == "/webhooks/wh-1":
			active := false
			json.NewEncoder(w).Encode(WebhookEndpoint{ID: "wh-1", IsActive: active})
		case r.Method == http.MethodDelete && r.URL.Path == "/webhooks/wh-1":
			w.WriteHeader(http.StatusNoContent)
		default:
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
	})

	webhooks, err := client.ListWebhooks(ctx)
	if err != nil || len(webhooks) != 1 {
		t.Fatalf("ListWebhooks: %+v, %v", webhooks, err)
	}

	created, err := client.CreateWebhook(ctx, CreateWebhookInput{
		URL:    "https://example.com/hook",
		Events: []WebhookEventType{EventNotificationDelivered},
	})
	if err != nil || created.Secret != "whsec_shown_once" {
		t.Fatalf("CreateWebhook: %+v, %v", created, err)
	}

	inactive := false
	updated, err := client.UpdateWebhook(ctx, "wh-1", UpdateWebhookInput{IsActive: &inactive})
	if err != nil || updated.IsActive {
		t.Fatalf("UpdateWebhook: %+v, %v", updated, err)
	}

	if err := client.DeleteWebhook(ctx, "wh-1"); err != nil {
		t.Fatalf("DeleteWebhook: %v", err)
	}
}

func TestNewRequiresCredentials(t *testing.T) {
	if _, err := New(Config{}); err == nil {
		t.Fatal("expected error for missing credentials")
	}
	if _, err := New(Config{APIKey: "k"}); err == nil {
		t.Fatal("expected error for missing api secret")
	}
}

// TestConcurrentCallersShareOneTokenExchange proves the tokenMu
// hold-lock-across-I/O design in authenticatedToken (see the comment
// there): many callers racing to make their first authenticated request
// must still result in exactly one /auth/token exchange, not one per
// goroutine. Run with -race to also confirm no data race on the shared
// token cache itself.
func TestConcurrentCallersShareOneTokenExchange(t *testing.T) {
	ctx := context.Background()
	const concurrentCallers = 50

	// A small delay widens the race window: without correct
	// serialization, every goroutine would observe an empty cache and
	// start its own token exchange before the first one completes.
	_, client, tokenCallCount := newTestServer(t, func(w http.ResponseWriter, r *http.Request, token string) {
		time.Sleep(5 * time.Millisecond)
		json.NewEncoder(w).Encode([]ChannelConfig{})
	})

	var wg sync.WaitGroup
	errs := make(chan error, concurrentCallers)
	for i := 0; i < concurrentCallers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			if _, err := client.ListChannels(ctx); err != nil {
				errs <- err
			}
		}()
	}
	wg.Wait()
	close(errs)

	for err := range errs {
		t.Errorf("ListChannels: %v", err)
	}
	if got := tokenCallCount(); got != 1 {
		t.Fatalf("got %d /auth/token exchanges for %d concurrent callers, want exactly 1", got, concurrentCallers)
	}
}

File: sdks/typescript/test/client.test.ts
import { describe, expect, it } from "vitest";
import { NotifydClient } from "../src/client.js";
import { NotifydConfigError, NotifydRequestError } from "../src/errors.js";
import { newMockServer } from "./mockServer.js";

describe("token exchange and caching", () => {
  it("reuses the cached token across multiple calls", async () => {
    let channelsCalls = 0;
    const { client } = newMockServer((request) => {
      expect(request.path).toBe("/channels");
      channelsCalls += 1;
      expect(request.bearerToken).toBe("test-token-1");
      return { status: 200, body: [] };
    });

    await client.listChannels();
    await client.listChannels();

    expect(channelsCalls).toBe(2);
  });

  it("throws NotifydConfigError when credentials are missing", () => {
    expect(() => new NotifydClient({ apiKey: "", apiSecret: "" })).toThrow(NotifydConfigError);
    expect(() => new NotifydClient({ apiKey: "k", apiSecret: "" })).toThrow(NotifydConfigError);
  });
});

describe("refresh-once-on-401", () => {
  it("retries exactly once with a fresh token after a 401", async () => {
    const seenTokens: string[] = [];
    const { client } = newMockServer((request) => {
      seenTokens.push(request.bearerToken);
      if (request.bearerToken === "test-token-1") {
        return { status: 401, body: { error: "TOKEN_EXPIRED" } };
      }
      return { status: 200, body: [] };
    });

    await client.listChannels();

    expect(seenTokens).toEqual(["test-token-1", "test-token-2"]);
  });

  it("gives up after one retry instead of looping on persistent 401s", async () => {
    let attempts = 0;
    const { client } = newMockServer(() => {
      attempts += 1;
      return { status: 401, body: { error: "INVALID_TOKEN" } };
    });

    await expect(client.listChannels()).rejects.toThrow(NotifydRequestError);
    expect(attempts).toBe(2);
  });

  it("surfaces the error code from the final 401", async () => {
    const { client } = newMockServer(() => ({ status: 401, body: { error: "INVALID_TOKEN" } }));

    try {
      await client.listChannels();
      expect.unreachable("expected listChannels to throw");
    } catch (error) {
      expect(error).toBeInstanceOf(NotifydRequestError);
      expect((error as NotifydRequestError).statusCode).toBe(401);
      expect((error as NotifydRequestError).code).toBe("INVALID_TOKEN");
    }
  });
});

describe("notifications", () => {
  it("sends a notification", async () => {
    const { client, requests } = newMockServer((request) => {
      if (request.path === "/notifications/send") {
        return { status: 202, body: { id: "notif-1", status: "pending" } };
      }
      throw new Error(`unexpected request: ${request.method} ${request.path}`);
    });

    const notification = await client.send({ channelConfigId: "chan-1", body: "hello" });

    expect(notification.id).toBe("notif-1");
    expect(notification.status).toBe("pending");
    expect(requests[0].body).toEqual({
      channel_config_id: "chan-1",
      subject: undefined,
      body: "hello",
      metadata: undefined,
    });
  });

  it("exposes upgrade_url from a 429 QUOTA_EXCEEDED body", async () => {
    const { client } = newMockServer((request) => {
      expect(request.path).toBe("/notifications/send");
      return {
        status: 429,
        body: {
          error: "QUOTA_EXCEEDED",
          upgrade_url: "https://notifyd.fluxintek.com/account/upgrade",
        },
      };
    });

    try {
      await client.send({ channelConfigId: "chan-1", body: "hello" });
      expect.unreachable("expected send to throw");
    } catch (error) {
      expect(error).toBeInstanceOf(NotifydRequestError);
      const requestError = error as NotifydRequestError;
      expect(requestError.statusCode).toBe(429);
      expect(requestError.code).toBe("QUOTA_EXCEEDED");
      expect(requestError.body?.upgrade_url).toBe("https://notifyd.fluxintek.com/account/upgrade");
      expect(requestError.field("upgrade_url")).toBe("https://notifyd.fluxintek.com/account/upgrade");
    }
  });

  it("sends to multiple channels and surfaces partial errors", async () => {
    const { client } = newMockServer((request) => {
      expect(request.path).toBe("/notifications/send-multi");
      return {
        status: 202,
        body: {
          sent: [{ id: "notif-1" }],
          errors: ["chan-2: CHANNEL_NOT_FOUND"],
        },
      };
    });

    const result = await client.sendMulti([
      { channelConfigId: "chan-1", body: "a" },
      { channelConfigId: "chan-2", body: "b" },
    ]);

    expect(result.sent).toHaveLength(1);
    expect(result.errors).toEqual(["chan-2: CHANNEL_NOT_FOUND"]);
  });

  it("lists notifications with query filters", async () => {
    const { client, requests } = newMockServer((request) => {
      expect(request.path).toBe("/notifications");
      expect(request.searchParams.get("status")).toBe("delivered");
      return {
        status: 200,
        body: { data: [{ id: "notif-1", status: "delivered" }], total: 1, limit: 20, offset: 0 },
      };
    });

    const list = await client.listNotifications({ status: "delivered" });

    expect(list.total).toBe(1);
    expect(requests[0].searchParams.get("status")).toBe("delivered");
  });

  it("gets a notification by id", async () => {
    const { client } = newMockServer((request) => {
      expect(request.path).toBe("/notifications/notif-1");
      return { status: 200, body: { id: "notif-1", status: "delivered" } };
    });

    const notification = await client.getNotification("notif-1");

    expect(notification.status).toBe("delivered");
  });

  it("lists delivery attempts", async () => {
    const { client } = newMockServer((request) => {
      expect(request.path).toBe("/notifications/notif-1/attempts");
      return { status: 200, body: [{ id: "attempt-1", attempt_number: 1, status: "success" }] };
    });

    const attempts = await client.listAttempts("notif-1");

    expect(attempts).toHaveLength(1);
    expect(attempts[0].status).toBe("success");
  });
});

describe("channels", () => {
  it("supports the full CRUD lifecycle", async () => {
    const { client } = newMockServer((request) => {
      if (request.method === "POST" && request.path === "/channels") {
        return { status: 201, body: { id: "chan-1", channel: "telegram" } };
      }
      if (request.method === "GET" && request.path === "/channels/chan-1") {
        return { status: 200, body: { id: "chan-1", channel: "telegram" } };
      }
      if (request.method === "PATCH" && request.path === "/channels/chan-1") {
        return { status: 200, body: { id: "chan-1", channel: "telegram", is_active: false } };
      }
      if (request.method === "DELETE" && request.path === "/channels/chan-1") {
        return { status: 204 };
      }
      throw new Error(`unexpected request: ${request.method} ${request.path}`);
    });

    const created = await client.createChannel({
      channel: "telegram",
      name: "ops-alerts",
      config: { bot_token: "x", chat_id: "y" },
    });
    expect(created.id).toBe("chan-1");

    const fetched = await client.getChannel("chan-1");
    expect(fetched.id).toBe("chan-1");

    const updated = await client.updateChannel("chan-1", { is_active: false });
    expect(updated.is_active).toBe(false);

    await expect(client.deleteChannel("chan-1")).resolves.toBeUndefined();
  });
});

describe("api keys", () => {
  it("supports list, create, and revoke", async () => {
    const { client } = newMockServer((request) => {
      if (request.method === "GET" && request.path === "/keys") {
        return { status: 200, body: [{ id: "key-1", label: "ci" }] };
      }
      if (request.method === "POST" && request.path === "/keys") {
        return {
          status: 201,
          body: { key: { id: "key-2", label: "new" }, api_secret: "shown-once" },
        };
      }
      if (request.method === "DELETE" && request.path === "/keys/key-1") {
        return { status: 204 };
      }
      throw new Error(`unexpected request: ${request.method} ${request.path}`);
    });

    const keys = await client.listApiKeys();
    expect(keys).toHaveLength(1);

    const created = await client.createApiKey("new");
    expect(created.api_secret).toBe("shown-once");

    await expect(client.revokeApiKey("key-1")).resolves.toBeUndefined();
  });
});

describe("webhooks", () => {
  it("supports the full CRUD lifecycle", async () => {
    const { client } = newMockServer((request) => {
      if (request.method === "GET" && request.path === "/webhooks") {
        return { status: 200, body: [{ id: "wh-1" }] };
      }
      if (request.method === "POST" && request.path === "/webhooks") {
        return {
          status: 201,
          body: {
            id: "wh-2",
            url: "https://example.com/hook",
            events: ["notification.delivered"],
            secret: "whsec_shown_once",
          },
        };
      }
      if (request.method === "PUT" && request.path === "/webhooks/wh-1") {
        return { status: 200, body: { id: "wh-1", is_active: false } };
      }
      if (request.method === "DELETE" && request.path === "/webhooks/wh-1") {
        return { status: 204 };
      }
      throw new Error(`unexpected request: ${request.method} ${request.path}`);
    });

    const webhooks = await client.listWebhooks();
    expect(webhooks).toHaveLength(1);

    const created = await client.createWebhook({
      url: "https://example.com/hook",
      events: ["notification.delivered"],
    });
    expect(created.secret).toBe("whsec_shown_once");

    const updated = await client.updateWebhook("wh-1", { is_active: false });
    expect(updated.is_active).toBe(false);

    await expect(client.deleteWebhook("wh-1")).resolves.toBeUndefined();
  });
});
